### PR TITLE
Add route /crud to execute sql statements on html page

### PR DESCRIPTION
**Pull request by nikolockenvitz**
- `/crud` displays a html page where the statement can be entered and submitted, the result is displayed there too
- `/sql` is the route to exchange data for executing statements in json format, statement goes in, get's executed, result will be returned
- restricted to authorized users, only usernames `niko` and `manuel`
- see also issue #3

### CODE REVIEW

File: src/rcj_flask.py [hunk 1]
```diff
@@ -94,3 +94,72 @@ def get_runs():
 def get_runs_competition(competition):
     return jsonify({'runs': g.rcj.get_runs_competition(competition)})
 
+def api_v2_check_auth(request):
+    username = request['referee']['name']
+    password = request['referee']['auth']
```
Comment (mb) [inline src/rcj_flask.py]: Check if `'referee'` with the attributes `'name'` and `'auth'` exists before accessing them.

File: src/rcj_flask.py [hunk 2]
```diff
@@ -94,3 +94,72 @@ def get_runs():
...
+@app.route('/api/v2/sql', methods=['POST'])
+def sql():
+    if not request.is_json:
+        return jsonify({'error': 'not json'}), 400
+    try:
+        req = request.json
+    except ValueError as e:
+        return jsonify({'error': str(e)}), 400
+
+    if(not api_v2_check_auth(req)):
```
Comment (mb) [inline src/rcj_flask.py]: Nitpick: don't use c-like brackets around if-statements.

File: src/rcj_flask.py [hunk 3]
```diff
@@ -94,3 +94,72 @@ def get_runs():
...
+    return send_from_directory('../public', 'main.js')
+
+@app.route('/api/v2/sql', methods=['POST'])
+def sql():
+    if not request.is_json:
+        return jsonify({'error': 'not json'}), 400
+    try:
+        req = request.json
+    except ValueError as e:
+        return jsonify({'error': str(e)}), 400
```
Comment (mb) [inline src/rcj_flask.py]: We already checked if it is json. No need for the try-except clause here.
I think we can also delete the try-except clause above in the "/api/v1/submit_run"-function.

**Review by mb — COMMENTED** [review]
Thanks. Looks good.

**Comment by nikolockenvitz** [comment]
You're absolutely right. I updated everything accordingly.

File: src/rcj_flask.py (hunk 1 above)
Comment (nikolockenvitz) [reply src/rcj_flask.py]: https://github.com/rcjberlin/rcj-server/blob/e41d2bed48182bd54b1cc712eea0f929a2f9f6d7/src/rcj_flask.py#L94-L100

**Comment by nikolockenvitz** [comment]
The idea was that if multiple statements get executed and one of them fails, all should fail (in case they depend on each other or so; first the problem should be fixed)
https://github.com/rcjberlin/rcj-server/blob/e41d2bed48182bd54b1cc712eea0f929a2f9f6d7/src/db.py#L29-L39
So in the except-block it returns without committing. Will these still be committed if the `self.db.commit()` runs a later time (e.g. right after unsuccessfully running someone else submits a run)? Do we need something like `BEGIN` and `ROLLBACK` here?

**Comment by nikolockenvitz** [comment]
> 
> 
> The idea was that if multiple statements get executed and one of them fails, all should fail (in case they depend on each other or so; first the problem should be fixed)
> 
> https://github.com/rcjberlin/rcj-server/blob/e41d2bed48182bd54b1cc712eea0f929a2f9f6d7/src/db.py#L29-L39
> 
> So in the except-block it returns without committing. Will these still be committed if the `self.db.commit()` runs a later time (e.g. right after unsuccessfully running someone else submits a run)? Do we need something like `BEGIN` and `ROLLBACK` here?

It works as desired due to closing the database connection after each request.

https://github.com/rcjberlin/rcj-server/blob/e41d2bed48182bd54b1cc712eea0f929a2f9f6d7/src/rcj_flask.py#L39-L47

**Pull request merged**